File: cpp/078子集/078子集.cpp
/*
给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。

说明：解集不能包含重复的子集。

示例:

输入: nums = [1,2,3]
输出:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
*/

//解答：用dfs深搜做。每次递归时，都将当前的数组加入result。并用变量pos防止重复

#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector <int>> result;
        vector<int> temp;
        result.push_back(temp);
        add(result, temp, nums, 0);
        return result;
        
    }
    
    void add(vector<vector <int>> &result, vector<int>& temp, vector<int>& nums, int pos)
    {
        if(pos == nums.size())
            return;
        
        for(int i = pos; i < nums.size(); i++)
        {
            temp.push_back(nums[i]);
            result.push_back(temp);
            add(result, temp, nums, i + 1);
            temp.pop_back();
        }
    }
};

File: python/078子集/078子集.py
"""
给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。

说明：解集不能包含重复的子集。

示例:

输入: nums = [1,2,3]
输出:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
"""
# 解答：可以用位运算来做。n个元素的子集个数为2^n, 外层遍历2^n次。每次右移n次，与1进行位于，相当于判断第一层遍历的索引i的n位中有几位是1，对应位数是1的将位数对应
# 的下标的元素放入当前数组即可。实质是2进制数可以完美的表达n个元素有2^n个子集。有每位是否是1来表示是否加入数组

class Solution:
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        for i in range(0, 2 ** len(nums)):      # n个元素有2^n个子集
            index = i
            temp = []
            for j in range(0, len(nums)):       # 每个index有n位。比如{1,2,3}求子集，那么000相当于空集，001相当于{1}，010相当于{2}，111相当于{1,2,3}
                if index >> j & 1:              # 每次进行右移和与1进行位于，相当于判断哪一位是否是1
                    temp.append(nums[j])
            
            result.append(temp)
        
        return result